### PR TITLE
[fx_importer] Add support for 0D tensors

### PR DESCRIPTION
**Pull request by dan-garvey**
Adds an escape hatch from creating a DenseResourceElementsAttr for single value tensors into DenseElementsAttr.

Addresses #398

### CODE REVIEW

File: core/shark_turbine/importers/fx_importer.py [hunk 1]
```diff
@@ -925,15 +928,21 @@ def _make_vtensor_literal_op(
...
-        bytes_view = memoryview(np_tensor)
-        tensor_type = create_mlir_tensor_type(tensor)
-        shape_desc = "_".join([str(d) for d in tensor.shape])
-        blob_name = f"torch_tensor_{shape_desc}_{str(tensor.dtype)}"
-        elements_attr = DenseResourceElementsAttr.get_from_buffer(
-            bytes_view,
-            blob_name,
-            tensor_type,
-        )
+        # DenseResourceElementsAttr creation doesnt support rank 0 tensors, so we use DenseElementsAttr instead.
```
Comment (stellaraccident) [inline core/shark_turbine/importers/fx_importer.py:931]: The comment should start with: one element constants are more optimizable as splat DenseElementsAttr. DenseResourceElementsAttr does not support splats, so don't use it for that case. In addition, at the time of writing, it has bugs with handling 0d tensors.

File: core/shark_turbine/importers/fx_importer.py [hunk 2]
```diff
@@ -925,15 +928,21 @@ def _make_vtensor_literal_op(
...
+        if np_tensor.size == 1:
+            elements_attr = DenseElementsAttr.get(
+                type=TORCH_DTYPE_TO_MLIR_TYPE[tensor.dtype](), array=np_tensor
```
Comment (stellaraccident) [inline core/shark_turbine/importers/fx_importer.py:934]: Can you also pass the shape= param? Keeps it from making dodgy local decisions and is marginally more robust.

Comment (stellaraccident) [inline core/shark_turbine/importers/fx_importer.py:934]: Wrap in try...except KeyError and raise a descriptive error message.

**Review by stellaraccident — APPROVED** [review]
Thanks. A few comments to address before landing.

**Comment by stellaraccident** [comment]
I'll need to sync this to torch-mlir once this lands. We're halfway through using that one as sot so need to port patches manually until it resolves. Can you leave the issue open and assign to me to port so I don't forget?

File: core/shark_turbine/importers/fx_importer.py [hunk 3]
```diff
@@ -928,10 +928,15 @@ def _make_vtensor_literal_op(
...
+        # one element constants are more optimizable as splat DenseElementsAttr. DenseResourceElementsAttr does not support splats, so don't use it for that case. In addition, at the time of writing, it has bugs with handling 0d tensors.
         if np_tensor.size == 1:
+            try:
+                dtype = tensor.dtype
+                element_type = TORCH_DTYPE_TO_MLIR_TYPE[dtype]()
+            except KeyError:
+                raise TypeError(f"Could not map Torch dtype {dtype} to an MLIR type")
             elements_attr = DenseElementsAttr.get(
-                type=TORCH_DTYPE_TO_MLIR_TYPE[tensor.dtype](), array=np_tensor
+                type=element_type, array=np_tensor, shape=[]
```
Comment (stellaraccident) [inline core/shark_turbine/importers/fx_importer.py:939]: Shape param is wrong for 1d, 1 element. Get it from np_tensor.shape

File: core/shark_turbine/importers/fx_importer.py [hunk 4]
```diff
@@ -928,10 +928,15 @@ def _make_vtensor_literal_op(
         # buffer is via the indirection: Tensor -> list -> numpy array. This allows us to create a vtensor literal as
         # desired, but also limits which data types we can support in this function (see TORCH_DTYPE_TO_NPY_TYPE above)
         np_tensor = np.array(tensor.tolist()).astype(npy_dtype)
-        # DenseResourceElementsAttr creation doesnt support rank 0 tensors, so we use DenseElementsAttr instead.
+        # one element constants are more optimizable as splat DenseElementsAttr. DenseResourceElementsAttr does not support splats, so don't use it for that case. In addition, at the time of writing, it has bugs with handling 0d tensors.
```
Comment (stellaraccident) [inline core/shark_turbine/importers/fx_importer.py:931]: Capitalize first letter.

**Review by stellaraccident — APPROVED** [review]
Couple of comments

File: core/shark_turbine/importers/fx_importer.py (hunk 3 above)
Comment (dan-garvey) [reply core/shark_turbine/importers/fx_importer.py:941]: oof, sounds like a missing test case

Comment (stellaraccident) [reply core/shark_turbine/importers/fx_importer.py:941]: Yeah, add tests for shape [1].

**Pull request merged**